File: LibraryManagementSystem/src/main/java/com/dat/LibraryManagementSystem/service/GenreService.java
package com.dat.LibraryManagementSystem.service;
import com.dat.LibraryManagementSystem.exception.GenreException;
import com.dat.LibraryManagementSystem.payload.dto.GenreDTO;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;




public interface GenreService {
    GenreDTO createGenre (GenreDTO genre);

    List<GenreDTO> getAllGenres();

    GenreDTO getGenreById(Long genreId) throws GenreException;

    GenreDTO updateGenre(Long genreId, GenreDTO genreDTO) throws GenreException;

    void deleteGenre(Long genreId) throws GenreException;

    void hardDeleteGenre(Long genreId) throws GenreException;



    List<GenreDTO> getAllActiveGenresWithSubGenres();

    List<GenreDTO> getSubGenres(Long genreId);

//    List<GenreDTO> getTopLevelGenresWithSubGenres();

    List<GenreDTO> getTopLevelGenres();

    // Page<GenreDTO> searchGenres(String searchTerm, Pageable pageable);

    long getTotalActiveGenres();

    long getBookCountByGenres(Long genreId);

}

File: frontend/library-frontend/src/services/genreService.js
import api from "./api";

export const genreService = {
  searchGenres: (params) => api.get("/genres", { params }),
  createGenre: (data) => api.post("/genres/create", data),  // sửa endpoint
  updateGenre: (id, data) => api.put(`/genres/${id}`, data),
  deleteGenre: (id) => api.delete(`/genres/${id}`),
  hardDeleteGenre: (id) => api.delete(`/genres/${id}/hard`),
  getAllGenres: () => api.get("/genres/get"),
  getTopLevelGenres: () => api.get("/genres/top-level"),
  getTopBorrowedGenres: (limit = 5) => api.get("/genres/top-borrowed", { params: { limit } }),
  getTotalActiveGenres: () => api.get("/genres/count"),
  getBookCountByGenre: (id) => api.get(`/genres/${id}/book-count`),
  getGenreById: (id) => api.get(`/genres/${id}`),
};
